engines/dfc_provider/app/services/dfc_product_type_factory.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require 'singleton'
+
+class DfcProductTypeFactory
+  include Singleton
+
+  def self.for(dfc_id)
+    instance.for(dfc_id)
+  end
+
+  def initialize
+    @product_types = {}
+
+    populate_product_types
+  end
+
+  def for(dfc_id)
+    return nil if @product_types[dfc_id].nil?
+
+    call_dfc_product_type(@product_types[dfc_id])
+  end
+
+  private
+
+  def populate_product_types
+    DfcLoader.connector.PRODUCT_TYPES.topConcepts.each do |product_type|
+      stack = []
+      record_type(stack, product_type.to_s)
+    end
+  end
+
+  def record_type(stack, product_type)
+    name = product_type.to_s
+    current_stack = stack.dup.push(name)
+
+    type = call_dfc_product_type(current_stack)
+
+    id = type.semanticId
+    @product_types[id] = current_stack
```

Comment:
Why don't you store the type directly? Then you don't need to resolve the stack when looking up types by id.

Reply:
I am not sure to be honest, it makes so much sense now that you are pointing it out :smile: